In [184]:
import pandas as pd
import numpy as np
import time
import datetime
from datetime import datetime
import sklearn
from sklearn.linear_model import LinearRegression
import matplotlib
import matplotlib.pyplot as plt
import math

In [185]:
df = pd.read_csv("bmw_pricing_v2.csv")

In [186]:
df.head()

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
0   BMW    118  140411.0     100.0     2012-02-01        diesel   black   
1   BMW     M4   13929.0     317.0     2016-04-01        petrol    grey   
2   BMW    320  183297.0     120.0     2012-04-01        diesel   white   
3   BMW    420  128035.0     135.0     2014-07-01        diesel     red   
4   BMW    425   97097.0     160.0     2014-12-01        diesel  silver   

    tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0  convertible              True               True          False   
1  convertible              True               True          False   
2  convertible             False              False          False   
3  convertible              True               True          False   
4  convertible              True               True          False   

  asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
0                       False                 True      True  True   
1                       False                False      True  True   
2                       False                 True     False  True   
3                       False                 True      True  True   
4                       False                False      True  True   

  alerta_lim_velocidad   precio fecha_venta  
0                False  11300.0  2018-01-01  
1                 True  69700.0  2018-02-01  
2                False  10200.0  2018-02-01  
3                 True  25100.0  2018-02-01  
4                 True  33400.0  2018-04-01

In [187]:
df["km"][459]

nan

In [188]:
df.shape

(4843, 18)

In [189]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        4841 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               4842 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4831 non-null   object 
 7   tipo_coche                   4834 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4841 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  4839 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

#### 1) Hacer un listado de los vehículos agrupados por el modelo y tipo de gasolina.

In [190]:
df['modelo'].value_counts().sum()

4840

In [191]:
# Tenemos 4840 modelos de coches
df.shape

(4843, 18)

In [192]:
# Vemos que tenemos 4843 filas
4843 - 4840

3

In [193]:
df["tipo_gasolina"].value_counts()

diesel           4631
petrol            191
hybrid_petrol       8
Diesel              5
electro             3
Name: tipo_gasolina, dtype: int64

Vamos a juntar las variables "Diesel" y "diesel" para agruparlas en la misma variable

In [194]:
df["tipo_gasolina"].replace({"diesel":"Diesel"}, inplace=True)

Ahora ya lo tenemos todo agrupado en "Diesel"

In [195]:
df["tipo_gasolina"].value_counts()

Diesel           4636
petrol            191
hybrid_petrol       8
electro             3
Name: tipo_gasolina, dtype: int64

In [196]:
# Vemos que tenemos 5 tipos de gasolina
# Vamos a hacer el group by por "modelo" y "tipo de gasolina"
df.groupby("modelo")["tipo_gasolina"].value_counts().to_frame()

tipo_gasolina
modelo         tipo_gasolina               
 Active Tourer Diesel                     1
114            Diesel                    10
               petrol                     5
116            Diesel                   339
               petrol                    19
...                                     ...
X6 M           petrol                     1
Z4             petrol                     6
i3             electro                    3
               hybrid_petrol              3
i8             hybrid_petrol              1

[106 rows x 1 columns]

#### 2) ¿Cuantos vehículos tienen potencia superior a los 150, que sean de tipo de coche convertible y tengan el volante regulable?

In [197]:
df.columns

Index(['marca', 'modelo', 'km', 'potencia', 'fecha_registro', 'tipo_gasolina',
       'color', 'tipo_coche', 'volante_regulable', 'aire_acondicionado',
       'camara_trasera', 'asientos_traseros_plegables', 'elevalunas_electrico',
       'bluetooth', 'gps', 'alerta_lim_velocidad', 'precio', 'fecha_venta'],
      dtype='object')

In [198]:
df["potencia"]

0       100.0
1       317.0
2       120.0
3       135.0
4       160.0
        ...  
4838    110.0
4839    100.0
4840    110.0
4841    110.0
4842    160.0
Name: potencia, Length: 4843, dtype: float64

In [199]:
df["tipo_coche"]

0       convertible
1       convertible
2       convertible
3       convertible
4       convertible
           ...     
4838            van
4839            van
4840            van
4841            van
4842            NaN
Name: tipo_coche, Length: 4843, dtype: object

In [200]:
df["volante_regulable"]

0        True
1        True
2       False
3        True
4        True
        ...  
4838    False
4839    False
4840    False
4841     True
4842     True
Name: volante_regulable, Length: 4843, dtype: object

In [201]:
df[(df["potencia"]> 150.0) & (df["tipo_coche"] == "convertible") & (df["volante_regulable"]==True)]

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
1    BMW     M4   13929.0     317.0     2016-04-01        petrol    grey   
4    BMW    425   97097.0     160.0     2014-12-01        Diesel  silver   
5    BMW    335  152352.0     225.0     2011-05-01        petrol   black   
17   BMW    650   24521.0     270.0     2008-01-01        petrol    grey   
24   BMW     Z4   43662.0     180.0     2013-05-01        petrol  orange   
26   BMW    430  113678.0     190.0     2014-07-01        Diesel   black   
37   BMW    650   24452.0     270.0     2008-01-01        petrol    grey   
41   BMW     M3  270907.0     236.0     1996-08-01        petrol   black   
42   BMW    330  142483.0     180.0     2012-01-01        Diesel    blue   

     tipo_coche volante_regulable aire_acondicionado camara_trasera  \
1   convertible              True               True          False   
4   convertible              True               True          False   
5   convertible              True               True          False   
17  convertible              True              False          False   
24  convertible              True              False          False   
26  convertible              True               True          False   
37  convertible              True              False          False   
41  convertible              True              False          False   
42  convertible              True               True          False   

   asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
1                        False                False      True  True   
4                        False                False      True  True   
5                        False                 True      True  True   
17                       False                False     False  True   
24                       False                 True     False  True   
26                       False                 True      True  True   
37                       False                False     False  True   
41                       False                False      True  True   
42                       False                 True     False  True   

   alerta_lim_velocidad   precio fecha_venta  
1                  True  69700.0  2018-02-01  
4                  True  33400.0  2018-04-01  
5                  True  17100.0  2018-02-01  
17                False   9200.0  2018-09-01  
24                False  20800.0  2018-05-01  
26                 True  30100.0  2018-05-01  
37                False   6800.0  2018-07-01  
41                False  12100.0  2018-08-01  
42                 True  15800.0  2007-08-01

In [202]:
df[(df["potencia"]> 150.0) & (df["tipo_coche"] == "convertible") & (df["volante_regulable"]==True)]

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
1    BMW     M4   13929.0     317.0     2016-04-01        petrol    grey   
4    BMW    425   97097.0     160.0     2014-12-01        Diesel  silver   
5    BMW    335  152352.0     225.0     2011-05-01        petrol   black   
17   BMW    650   24521.0     270.0     2008-01-01        petrol    grey   
24   BMW     Z4   43662.0     180.0     2013-05-01        petrol  orange   
26   BMW    430  113678.0     190.0     2014-07-01        Diesel   black   
37   BMW    650   24452.0     270.0     2008-01-01        petrol    grey   
41   BMW     M3  270907.0     236.0     1996-08-01        petrol   black   
42   BMW    330  142483.0     180.0     2012-01-01        Diesel    blue   

     tipo_coche volante_regulable aire_acondicionado camara_trasera  \
1   convertible              True               True          False   
4   convertible              True               True          False   
5   convertible              True               True          False   
17  convertible              True              False          False   
24  convertible              True              False          False   
26  convertible              True               True          False   
37  convertible              True              False          False   
41  convertible              True              False          False   
42  convertible              True               True          False   

   asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
1                        False                False      True  True   
4                        False                False      True  True   
5                        False                 True      True  True   
17                       False                False     False  True   
24                       False                 True     False  True   
26                       False                 True      True  True   
37                       False                False     False  True   
41                       False                False      True  True   
42                       False                 True     False  True   

   alerta_lim_velocidad   precio fecha_venta  
1                  True  69700.0  2018-02-01  
4                  True  33400.0  2018-04-01  
5                  True  17100.0  2018-02-01  
17                False   9200.0  2018-09-01  
24                False  20800.0  2018-05-01  
26                 True  30100.0  2018-05-01  
37                False   6800.0  2018-07-01  
41                False  12100.0  2018-08-01  
42                 True  15800.0  2007-08-01

In [203]:
len(df[(df["potencia"]> 150.0) & (df["tipo_coche"] == "convertible") & (df["volante_regulable"]==True)])

9

#### 3)¿Podrías indicar si hay algún caso que la fecha de registro sea superior a la fecha de venta? De ser así indicar el precio medio encontrado?

In [204]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        4841 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               4842 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4831 non-null   object 
 7   tipo_coche                   4834 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4841 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  4839 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

Vemos que las fechas no esta pasadas a "datetime", vamos a convertirlas.

In [205]:
df['fecha_registro'] = pd.to_datetime(df['fecha_registro'], format = '%Y.%m.%d')

In [206]:
df['fecha_venta'] = pd.to_datetime(df['fecha_venta'], format = '%Y.%m.%d')

In [207]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4841 non-null   float64       
 3   potencia                     4842 non-null   float64       
 4   fecha_registro               4842 non-null   datetime64[ns]
 5   tipo_gasolina                4838 non-null   object        
 6   color                        4831 non-null   object        
 7   tipo_coche                   4834 non-null   object        
 8   volante_regulable            4839 non-null   object        
 9   aire_acondicionado           4841 non-null   object        
 10  camara_trasera               4841 non-null   object        
 11  asientos_traseros_plegables  4839 non-null 

In [208]:
df[(df["fecha_registro"] > df["fecha_venta"])]

marca modelo        km  potencia fecha_registro tipo_gasolina  color  \
42    BMW    330  142483.0     180.0     2012-01-01        Diesel   blue   
100   BMW    320  342768.0     135.0     2011-12-01        Diesel  black   
180   BMW    320  181052.0     120.0     2014-10-01        Diesel  black   
256   BMW    316  201102.0      85.0     2013-12-01        Diesel  black   

      tipo_coche volante_regulable aire_acondicionado camara_trasera  \
42   convertible              True               True          False   
100        coupe             False               True          False   
180       estate             False               True          False   
256       estate              True              False          False   

    asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
42                        False                 True     False  True   
100                       False                 True     False  True   
180                       False                False     False  True   
256                       False                False     False  True   

    alerta_lim_velocidad   precio fecha_venta  
42                  True  15800.0  2007-08-01  
100                False   8200.0  2010-04-01  
180                 True  11500.0  2009-04-01  
256                False   8800.0  2008-06-01

In [209]:
df[(df["fecha_registro"] > df["fecha_venta"])]["precio"].describe()

count        4.000000
mean     11075.000000
std       3461.574786
min       8200.000000
25%       8650.000000
50%      10150.000000
75%      12575.000000
max      15800.000000
Name: precio, dtype: float64

El precio medio es de 11075

Indicar el precio medio de los vehículos con aire acondicionado.

In [210]:
df.columns

Index(['marca', 'modelo', 'km', 'potencia', 'fecha_registro', 'tipo_gasolina',
       'color', 'tipo_coche', 'volante_regulable', 'aire_acondicionado',
       'camara_trasera', 'asientos_traseros_plegables', 'elevalunas_electrico',
       'bluetooth', 'gps', 'alerta_lim_velocidad', 'precio', 'fecha_venta'],
      dtype='object')

In [211]:
df[(df["aire_acondicionado"]==True)]

marca             modelo        km  potencia fecha_registro  \
0      BMW                118  140411.0     100.0     2012-02-01   
1      BMW                 M4   13929.0     317.0     2016-04-01   
3      BMW                420  128035.0     135.0     2014-07-01   
4      BMW                425   97097.0     160.0     2014-12-01   
5      BMW                335  152352.0     225.0     2011-05-01   
...    ...                ...       ...       ...            ...   
4838   BMW    218 Gran Tourer   39743.0     110.0     2015-10-01   
4839   BMW  218 Active Tourer   49832.0     100.0     2015-06-01   
4840   BMW    218 Gran Tourer   19633.0     110.0     2015-10-01   
4841   BMW  218 Active Tourer   27920.0     110.0     2016-04-01   
4842   BMW                525  195840.0     160.0     2013-01-01   

     tipo_gasolina   color   tipo_coche volante_regulable aire_acondicionado  \
0           Diesel   black  convertible              True               True   
1           petrol    grey  convertible              True               True   
3           Diesel     red  convertible              True               True   
4           Diesel  silver  convertible              True               True   
5           petrol   black  convertible              True               True   
...            ...     ...          ...               ...                ...   
4838        Diesel   black          van             False               True   
4839        Diesel    grey          van             False               True   
4840        Diesel    grey          van             False               True   
4841        Diesel   brown          van              True               True   
4842        Diesel    grey          NaN              True               True   

     camara_trasera asientos_traseros_plegables elevalunas_electrico  \
0             False                       False                 True   
1             False                       False                False   
3             False                       False                 True   
4             False                       False                False   
5             False                       False                 True   
...             ...                         ...                  ...   
4838          False                       False                False   
4839          False                       False                False   
4840          False                       False                False   
4841          False                       False                False   
4842          False                       False                 True   

     bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
0         True  True                False  11300.0  2018-01-01  
1         True  True                 True  69700.0  2018-02-01  
3         True  True                 True  25100.0  2018-02-01  
4         True  True                 True  33400.0  2018-04-01  
5         True  True                 True  17100.0  2018-02-01  
...        ...   ...                  ...      ...         ...  
4838     False  True                False  14600.0  2018-08-01  
4839     False  True                 True  17500.0  2018-08-01  
4840     False  True                 True  17000.0  2018-09-01  
4841     False  True                 True  22700.0  2018-09-01  
4842     False  True                 True  15500.0  2018-09-01  

[3837 rows x 18 columns]

In [212]:
df[(df["aire_acondicionado"]==True)]["precio"].mean()

16998.12108559499

In [213]:
round(df[(df["aire_acondicionado"]==True)]["precio"].mean(),2)

16998.12

#### 4) ¿Existe alguna diferencia significativa entre el color del vehículos y el precio? Agrupar los vehículos por su color y su precio medio para responder la pregunta.

In [214]:
df.columns

Index(['marca', 'modelo', 'km', 'potencia', 'fecha_registro', 'tipo_gasolina',
       'color', 'tipo_coche', 'volante_regulable', 'aire_acondicionado',
       'camara_trasera', 'asientos_traseros_plegables', 'elevalunas_electrico',
       'bluetooth', 'gps', 'alerta_lim_velocidad', 'precio', 'fecha_venta'],
      dtype='object')

In [215]:
df.groupby("color")["precio"].mean().sort_values(ascending = False).to_frame()

precio
color               
orange  18866.666667
white   17314.365672
red     16568.627451
black   16178.479460
beige   15817.073171
grey    15619.914530
brown   15367.741935
blue    15137.784091
silver  14826.299694
green    7200.000000

In [216]:
# Redondeamos para no tener tantos decimales
round(df.groupby("color")["precio"].mean().sort_values(ascending = False).to_frame(),2)

precio
color           
orange  18866.67
white   17314.37
red     16568.63
black   16178.48
beige   15817.07
grey    15619.91
brown   15367.74
blue    15137.78
silver  14826.30
green    7200.00

Tras esto, se pide realizar un modelo predictivo en Python mediante una regresión lineal, incluyendo la exploración (al menos las
visualizaciones de evolución temporal del precio y el número de vehículos vendidos por modelo), el preprocesamiento de los datos y
la estimación mediante una regresión lineal (opcionalmente, explorar otros algoritmos)

### EDA


In [217]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4841 non-null   float64       
 3   potencia                     4842 non-null   float64       
 4   fecha_registro               4842 non-null   datetime64[ns]
 5   tipo_gasolina                4838 non-null   object        
 6   color                        4831 non-null   object        
 7   tipo_coche                   4834 non-null   object        
 8   volante_regulable            4839 non-null   object        
 9   aire_acondicionado           4841 non-null   object        
 10  camara_trasera               4841 non-null   object        
 11  asientos_traseros_plegables  4839 non-null 

In [218]:
df.shape

(4843, 18)

In [219]:
df.isnull().sum().sum()

62

In [220]:
df.head()

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
0   BMW    118  140411.0     100.0     2012-02-01        Diesel   black   
1   BMW     M4   13929.0     317.0     2016-04-01        petrol    grey   
2   BMW    320  183297.0     120.0     2012-04-01        Diesel   white   
3   BMW    420  128035.0     135.0     2014-07-01        Diesel     red   
4   BMW    425   97097.0     160.0     2014-12-01        Diesel  silver   

    tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0  convertible              True               True          False   
1  convertible              True               True          False   
2  convertible             False              False          False   
3  convertible              True               True          False   
4  convertible              True               True          False   

  asientos_traseros_plegables elevalunas_electrico bluetooth   gps  \
0                       False                 True      True  True   
1                       False                False      True  True   
2                       False                 True     False  True   
3                       False                 True      True  True   
4                       False                False      True  True   

  alerta_lim_velocidad   precio fecha_venta  
0                False  11300.0  2018-01-01  
1                 True  69700.0  2018-02-01  
2                False  10200.0  2018-02-01  
3                 True  25100.0  2018-02-01  
4                 True  33400.0  2018-04-01

In [221]:
df.set_index("fecha_registro", inplace=True)
df.head()

marca modelo        km  potencia tipo_gasolina   color  \
fecha_registro                                                          
2012-02-01       BMW    118  140411.0     100.0        Diesel   black   
2016-04-01       BMW     M4   13929.0     317.0        petrol    grey   
2012-04-01       BMW    320  183297.0     120.0        Diesel   white   
2014-07-01       BMW    420  128035.0     135.0        Diesel     red   
2014-12-01       BMW    425   97097.0     160.0        Diesel  silver   

                 tipo_coche volante_regulable aire_acondicionado  \
fecha_registro                                                     
2012-02-01      convertible              True               True   
2016-04-01      convertible              True               True   
2012-04-01      convertible             False              False   
2014-07-01      convertible              True               True   
2014-12-01      convertible              True               True   

               camara_trasera asientos_traseros_plegables  \
fecha_registro                                              
2012-02-01              False                       False   
2016-04-01              False                       False   
2012-04-01              False                       False   
2014-07-01              False                       False   
2014-12-01              False                       False   

               elevalunas_electrico bluetooth   gps alerta_lim_velocidad  \
fecha_registro                                                             
2012-02-01                     True      True  True                False   
2016-04-01                    False      True  True                 True   
2012-04-01                     True     False  True                False   
2014-07-01                     True      True  True                 True   
2014-12-01                    False      True  True                 True   

                 precio fecha_venta  
fecha_registro                       
2012-02-01      11300.0  2018-01-01  
2016-04-01      69700.0  2018-02-01  
2012-04-01      10200.0  2018-02-01  
2014-07-01      25100.0  2018-02-01  
2014-12-01      33400.0  2018-04-01

In [222]:
df.isnull().sum()

marca                           2
modelo                          3
km                              2
potencia                        1
tipo_gasolina                   5
color                          12
tipo_coche                      9
volante_regulable               4
aire_acondicionado              2
camara_trasera                  2
asientos_traseros_plegables     4
elevalunas_electrico            2
bluetooth                       4
gps                             0
alerta_lim_velocidad            2
precio                          6
fecha_venta                     1
dtype: int64

Vemos que tenemos muchos nulos que hemos de imputar... Pero primero vamos analizar qué tipo de variables tenemos.

In [223]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4843 entries, 2012-02-01 to 2013-01-01
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4841 non-null   float64       
 3   potencia                     4842 non-null   float64       
 4   tipo_gasolina                4838 non-null   object        
 5   color                        4831 non-null   object        
 6   tipo_coche                   4834 non-null   object        
 7   volante_regulable            4839 non-null   object        
 8   aire_acondicionado           4841 non-null   object        
 9   camara_trasera               4841 non-null   object        
 10  asientos_traseros_plegables  4839 non-null   object        
 11  elevalunas_electrico     

Lista de booleanas que hemos de convertir a **bool**

1. "volante_regulable"
2. "aire_acondicionado"
3. "camara_trasera"
4. "asientos_traseros_plegables"
5. "elevalunas_electrico"
5. "bluetooth"
6. "alerta_lim_velocidad"

In [224]:
# Vamos a checkear las variables booleanas
df["volante_regulable"].describe()

count     4839
unique       2
top       True
freq      2662
Name: volante_regulable, dtype: object

In [225]:
df["volante_regulable"] = df["volante_regulable"].astype(bool)

In [226]:
df["volante_regulable"].dtype

dtype('bool')

In [227]:
lista_booleanas = ["volante_regulable", "aire_acondicionado", "camara_trasera", "asientos_traseros_plegables", "elevalunas_electrico", "bluetooth", "alerta_lim_velocidad"]

In [228]:
for item in lista_booleanas:
    if item == item:
        df[item] = df[item].astype(bool)

In [229]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4843 entries, 2012-02-01 to 2013-01-01
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4841 non-null   float64       
 3   potencia                     4842 non-null   float64       
 4   tipo_gasolina                4838 non-null   object        
 5   color                        4831 non-null   object        
 6   tipo_coche                   4834 non-null   object        
 7   volante_regulable            4843 non-null   bool          
 8   aire_acondicionado           4843 non-null   bool          
 9   camara_trasera               4843 non-null   bool          
 10  asientos_traseros_plegables  4843 non-null   bool          
 11  elevalunas_electrico     

Ahora vamos a imputar nulos a "mano"

In [230]:
df.isnull().sum().sum()

41

Empezamos por las numericas

In [231]:
df.select_dtypes(include = ['number']).isnull().sum()

km          2
potencia    1
precio      6
dtype: int64

In [232]:
list_numerical = df.select_dtypes(include='number')
list_numerical

km  potencia   precio
fecha_registro                             
2012-02-01      140411.0     100.0  11300.0
2016-04-01       13929.0     317.0  69700.0
2012-04-01      183297.0     120.0  10200.0
2014-07-01      128035.0     135.0  25100.0
2014-12-01       97097.0     160.0  33400.0
...                  ...       ...      ...
2015-10-01       39743.0     110.0  14600.0
2015-06-01       49832.0     100.0  17500.0
2015-10-01       19633.0     110.0  17000.0
2016-04-01       27920.0     110.0  22700.0
2013-01-01      195840.0     160.0  15500.0

[4843 rows x 3 columns]

In [235]:
for item in list_numerical:
   if item == item:
      df[item].fillna(df[item].mean, inplace = True)

Nos cargamos los decimales, ya que todos son 0 y no nos aportan ningún valor (al contrario, puede que en un futuro nos den problemas)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4843 entries, 2012-02-01 to 2013-01-01
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4843 non-null   object        
 3   potencia                     4843 non-null   object        
 4   tipo_gasolina                4838 non-null   object        
 5   color                        4831 non-null   object        
 6   tipo_coche                   4834 non-null   object        
 7   volante_regulable            4843 non-null   bool          
 8   aire_acondicionado           4843 non-null   bool          
 9   camara_trasera               4843 non-null   bool          
 10  asientos_traseros_plegables  4843 non-null   bool          
 11  elevalunas_electrico     

In [ ]:
df["km"]

fecha_registro
2012-02-01    140411.0
2016-04-01     13929.0
2012-04-01    183297.0
2014-07-01    128035.0
2014-12-01     97097.0
                ...   
2015-10-01     39743.0
2015-06-01     49832.0
2015-10-01     19633.0
2016-04-01     27920.0
2013-01-01    195840.0
Name: km, Length: 4843, dtype: object

In [ ]:
df["km"].dtype

dtype('O')

Vemos que al imputar los nulos la variable "km" la convierte a 'object'

In [ ]:
df["precio"]

fecha_registro
2012-02-01    11300.0
2016-04-01    69700.0
2012-04-01    10200.0
2014-07-01    25100.0
2014-12-01    33400.0
               ...   
2015-10-01    14600.0
2015-06-01    17500.0
2015-10-01    17000.0
2016-04-01    22700.0
2013-01-01    15500.0
Name: precio, Length: 4843, dtype: object

In [ ]:
df["precio"].fillna(value=df["precio"].mean, inplace=True)

In [ ]:
df["precio"].dtype

dtype('O')

In [ ]:
df["potencia"].fillna(value=df["potencia"].mean, inplace=True)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4843 entries, 2012-02-01 to 2013-01-01
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4843 non-null   object        
 3   potencia                     4843 non-null   object        
 4   tipo_gasolina                4838 non-null   object        
 5   color                        4831 non-null   object        
 6   tipo_coche                   4834 non-null   object        
 7   volante_regulable            4843 non-null   bool          
 8   aire_acondicionado           4843 non-null   bool          
 9   camara_trasera               4843 non-null   bool          
 10  asientos_traseros_plegables  4843 non-null   bool          
 11  elevalunas_electrico     

### 4.1 Evolución temporal del precio

Para poder hacer el análisis temporal, añadiremos las columnas de:
* AÑO
* MES
* DIA

In [ ]:
df["fecha_venta"]

fecha_registro
2012-02-01   2018-01-01
2016-04-01   2018-02-01
2012-04-01   2018-02-01
2014-07-01   2018-02-01
2014-12-01   2018-04-01
                ...    
2015-10-01   2018-08-01
2015-06-01   2018-08-01
2015-10-01   2018-09-01
2016-04-01   2018-09-01
2013-01-01   2018-09-01
Name: fecha_venta, Length: 4843, dtype: datetime64[ns]

In [ ]:
df["year"] = df["fecha_venta"].dt.year
df["month"] = df["fecha_venta"].dt.month
df["days"] = df["fecha_venta"].dt.days_in_month

In [ ]:
df.columns

Index(['marca', 'modelo', 'km', 'potencia', 'tipo_gasolina', 'color',
       'tipo_coche', 'volante_regulable', 'aire_acondicionado',
       'camara_trasera', 'asientos_traseros_plegables', 'elevalunas_electrico',
       'bluetooth', 'gps', 'alerta_lim_velocidad', 'precio', 'fecha_venta',
       'year', 'month', 'days'],
      dtype='object')

In [ ]:
df.head()

marca modelo        km potencia tipo_gasolina   color  \
fecha_registro                                                         
2012-02-01       BMW    118  140411.0    100.0        Diesel   black   
2016-04-01       BMW     M4   13929.0    317.0        petrol    grey   
2012-04-01       BMW    320  183297.0    120.0        Diesel   white   
2014-07-01       BMW    420  128035.0    135.0        Diesel     red   
2014-12-01       BMW    425   97097.0    160.0        Diesel  silver   

                 tipo_coche  volante_regulable  aire_acondicionado  \
fecha_registro                                                       
2012-02-01      convertible               True                True   
2016-04-01      convertible               True                True   
2012-04-01      convertible              False               False   
2014-07-01      convertible               True                True   
2014-12-01      convertible               True                True   

                camara_trasera  asientos_traseros_plegables  \
fecha_registro                                                
2012-02-01               False                        False   
2016-04-01               False                        False   
2012-04-01               False                        False   
2014-07-01               False                        False   
2014-12-01               False                        False   

                elevalunas_electrico  bluetooth   gps  alerta_lim_velocidad  \
fecha_registro                                                                
2012-02-01                      True       True  True                 False   
2016-04-01                     False       True  True                  True   
2012-04-01                      True      False  True                 False   
2014-07-01                      True       True  True                  True   
2014-12-01                     False       True  True                  True   

                 precio fecha_venta    year  month  days  
fecha_registro                                            
2012-02-01      11300.0  2018-01-01  2018.0    1.0  31.0  
2016-04-01      69700.0  2018-02-01  2018.0    2.0  28.0  
2012-04-01      10200.0  2018-02-01  2018.0    2.0  28.0  
2014-07-01      25100.0  2018-02-01  2018.0    2.0  28.0  
2014-12-01      33400.0  2018-04-01  2018.0    4.0  30.0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4843 entries, 2012-02-01 to 2013-01-01
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   marca                        4841 non-null   object        
 1   modelo                       4840 non-null   object        
 2   km                           4843 non-null   object        
 3   potencia                     4843 non-null   object        
 4   tipo_gasolina                4838 non-null   object        
 5   color                        4831 non-null   object        
 6   tipo_coche                   4834 non-null   object        
 7   volante_regulable            4843 non-null   bool          
 8   aire_acondicionado           4843 non-null   bool          
 9   camara_trasera               4843 non-null   bool          
 10  asientos_traseros_plegables  4843 non-null   bool          
 11  elevalunas_electrico     

In [ ]:
df.head()

marca modelo        km potencia tipo_gasolina   color  \
fecha_registro                                                         
2012-02-01       BMW    118  140411.0    100.0        Diesel   black   
2016-04-01       BMW     M4   13929.0    317.0        petrol    grey   
2012-04-01       BMW    320  183297.0    120.0        Diesel   white   
2014-07-01       BMW    420  128035.0    135.0        Diesel     red   
2014-12-01       BMW    425   97097.0    160.0        Diesel  silver   

                 tipo_coche  volante_regulable  aire_acondicionado  \
fecha_registro                                                       
2012-02-01      convertible               True                True   
2016-04-01      convertible               True                True   
2012-04-01      convertible              False               False   
2014-07-01      convertible               True                True   
2014-12-01      convertible               True                True   

                camara_trasera  asientos_traseros_plegables  \
fecha_registro                                                
2012-02-01               False                        False   
2016-04-01               False                        False   
2012-04-01               False                        False   
2014-07-01               False                        False   
2014-12-01               False                        False   

                elevalunas_electrico  bluetooth   gps  alerta_lim_velocidad  \
fecha_registro                                                                
2012-02-01                      True       True  True                 False   
2016-04-01                     False       True  True                  True   
2012-04-01                      True      False  True                 False   
2014-07-01                      True       True  True                  True   
2014-12-01                     False       True  True                  True   

                 precio fecha_venta    year  month  days  
fecha_registro                                            
2012-02-01      11300.0  2018-01-01  2018.0    1.0  31.0  
2016-04-01      69700.0  2018-02-01  2018.0    2.0  28.0  
2012-04-01      10200.0  2018-02-01  2018.0    2.0  28.0  
2014-07-01      25100.0  2018-02-01  2018.0    2.0  28.0  
2014-12-01      33400.0  2018-04-01  2018.0    4.0  30.0